src/solver/solution.rs

```diff
@@ -0,0 +1,115 @@
+use crate::fluent::FluentResource;
+use std::rc::Rc;
+
+#[derive(Debug)]
+pub struct Solution {
+    pub width: usize,
+    pub depth: usize,
+    pub candidate: Vec<usize>,
+    pub idx: usize,
+    pub dirty: bool,
+
+    pub cache: Vec<Vec<Option<Option<Rc<FluentResource>>>>>,
```

Comment:
I'm having difficulty wrapping my head around the nested Options here. Please add a comment :)